Source/WebCore/platform/audio/SharedAudioDestination.cpp

```diff
@@ -0,0 +1,249 @@
+    , m_configurationQueue { WorkQueue::create("SharedAudioDestinationAdapter configuration queue") }
+    , m_configurationSemaphore(0)
+{
+}
+
+void SharedAudioDestinationAdapter::addRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        ASSERT(!m_renderers.contains(renderer));
+        m_renderers.append(WTFMove(renderer));
+        configureRenderThread(WTFMove(completionHandler));
+    });
+}
+
+void SharedAudioDestinationAdapter::removeRenderer(SharedAudioDestination& renderer, CompletionHandler<void(bool)>&& completionHandler)
+{
+    m_configurationQueue->dispatch([this, renderer = RefPtr { &renderer }, completionHandler = WTFMove(completionHandler)] () mutable {
+        ASSERT(m_renderers.contains(renderer));
+        m_renderers.removeFirst(renderer);
+        configureRenderThread(WTFMove(completionHandler));
+    });
+}
+
+void SharedAudioDestinationAdapter::configureRenderThread(CompletionHandler<void(bool)>&& completionHandler)
+{
+    ASSERT(m_configurationQueue->isCurrent());
+    {
+        LockHolder locker { m_renderLock };
+        m_newRenderers = m_renderers;
+        m_needsConfiguration = true;
+    }
+
+    bool shouldStart = !m_started && !m_renderers.isEmpty();
+    bool shouldStop = m_started && m_renderers.isEmpty();
+
+    // The AudioDestination must be started for configuration to take place.
+    if (shouldStart) {
+        // Dispatching to the main thread is required for destinations
+        // which are subclasses of AudioDestinationResampler.
+        callOnMainThreadAndWait([this] {
+            m_destination->start(nullptr, [] (bool) { });
+        });
+        m_started = true;
+    }
+
+    m_configurationSemaphore.wait();
+
+    // The AudioDestination must not be stopped before configuration takes place.
+    if (shouldStop) {
+        // Dispatching to the main thread is required for destinations
+        // which are subclasses of AudioDestinationResampler.
+        callOnMainThreadAndWait([this] {
+            m_destination->stop();
+        });
+        m_started = false;
+    }
+
+    // Move the previously configured list of renderers to the MainThread for destruction:
+    Vector<RefPtr<SharedAudioDestination>> renderersToDispose;
+    {
+        LockHolder locker { m_renderLock };
+        renderersToDispose = std::exchange(m_newRenderers, { });
+    }
+
+    callOnMainThread([completionHandler = WTFMove(completionHandler), rendersToDispose = WTFMove(renderersToDispose)]() mutable {
+        UNUSED_PARAM(rendersToDispose);
+        completionHandler(true);
+    });
+}
+
+void SharedAudioDestinationAdapter::render(AudioBus* sourceBus, AudioBus* destinationBus, size_t numberOfFrames, const AudioIOPosition& outputPosition)
+{
+    if (m_renderLock.tryLock()) {
```

Comment:
Please use `Locker<Lock>::tryLock()` in new code. No manual locking / unlocking.

Reply:
Oh nice, will do.